feat: User-lists collaborators CRUD endpoints

jpmcb opened the pull request:
## Description

- Creates the following crud routes:
  - `/v1/lists/{id}/contributors` - [GET] Get paginated set of contributors in the list
  - `/v1/lists/{id}/contributors` - [POST] Add new contributor to an existing list
  - `/v1/lists/{id}/contributors` - [DELETE] Remove contributor from an existing list
  - `/v1/lists/contributors` - [GET] Get paginated set of contributors to display when creating a list (w/filters)
- Sets `"integer"` type for swagger doc creation for user-list entities (this was a doubley nested array in the swagger doc before this change):

![Screenshot 2023-09-06 at 4 12 13 PM](https://github.com/open-sauced/api/assets/23109390/ff0caf8d-acba-4f55-8842-967436ddeaa8)

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Related to https://github.com/open-sauced/api/pull/293

## Mobile & Desktop Screenshots/Recordings

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [x] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?



## [optional] What gif best describes this PR or how it makes you feel?

![](https://media.giphy.com/media/pZGDZwmxOtEEo/giphy.gif)

<!-- note: PRs with deleted sections will be marked invalid -->

<!--
  For Work In Progress Pull Requests, please use the Draft PR feature,
  see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.
  
  For a timely review/response, please avoid force-pushing additional
  commits if your PR already received reviews or comments.
  
  Before submitting a Pull Request, please ensure you've done the following:
  - 📖 Read the Open Sauced Contributing Guide: https://github.com/open-sauced/.github/blob/main/CONTRIBUTING.md.
  - 📖 Read the Open Sauced Code of Conduct: https://github.com/open-sauced/.github/blob/main/CODE_OF_CONDUCT.md.
  - 👷‍♀️ Create small PRs. In most cases, this will be possible.
  - ✅ Provide tests for your changes.
  - 📝 Use descriptive commit messages.
  - 📗 Update any related documentation and include any relevant screenshots.
-->

brandonroberts commented on src/user-lists/user-list.service.ts:
```diff
@@ -82,11 +82,38 @@ export class UserListService {
   }
+
+  async findAllContributors(pageOptionsDto: PageOptionsDto): Promise<PageDto<DbUserListContributor>> {
+    const queryBuilder = this.userListContributorRepository.createQueryBuilder("user_list_contributors");
```
I think this should be a different query to get a list of all available contributors/users. We have users for a specific list covered in the one below. 

Thoughts?

jpmcb replied:
Ah: so this really should be querying for all `users.id` since a `user_list_contributors.user_id` is the relationship to be made? The way it's written now, this will paginate all of the rows in `user_list_contributors`

jpmcb replied:
I think I'm still abit confused what this route is supposed to do. In my eyes, `/v1/lists/contributors [GET]` would return all contributors across all lists (what this route does right now).

But if we want all _possible_ contributors (i.e., all current users), wouldn't we just use the `/v1/users/search` endpoint?

brandonroberts replied:
We could use `/v1/users/search` but I'm not sure we would need the same filters there as for a contributor search. I don't think we need an endpoint to return all contributors across lists because each list is tied to the authenticated user.

brandonroberts commented on src/user-lists/user-list.service.ts:
```diff
@@ -82,11 +82,38 @@ export class UserListService {
     return this.userListContributorRepository.save(newUserListContributor);
   }
 
+  async deleteUserListContributor(userId: number) {
```
I think this should be a `string`, as the user list contributor record ID is a uuid

jpmcb replied:
Whoops, this should really be:

```ts
async deleteUserListContributor(id: number)
```

since the intent is the delete the contributor by its id

brandonroberts commented on src/user-lists/user-list.service.ts:
```diff
@@ -82,11 +82,38 @@ export class UserListService {
+  ): Promise<PageDto<DbUserListContributor>> {
+    const queryBuilder = this.userListContributorRepository.createQueryBuilder("user_list_contributors");
+
+    queryBuilder.where("user_list_contributors.list_id = :listId", { listId });
```
Same as above with joining the `users` table

jpmcb replied:
So we're not really returning "list contributors", but rather, we're finding the real `users` and getting those? Sounds like `user_list_contributors` is more of a "bridge" table in order to get `users`?

brandonroberts commented on src/user-lists/user-list.controller.ts:
```diff
@@ -102,4 +107,74 @@ export class UserListController {
+    return Promise.all(contributors);
+  }
+
+  @Delete("/:id/contributors")
```
We should be more specific here to only delete the individual record. The `:id` is the list ID

```suggestion
  @Delete("/:id/contributors/:userListContributorId")
```

jpmcb commented on src/user-lists/entities/user-list-contributor.entity.ts:
```diff
@@ -15,15 +16,15 @@ import {
-    example: 196,
-    type: "integer",
+    description: "User list contributor identifier",
+    example: "uuid-v4",
   })
+  @PrimaryColumn()
   @PrimaryGeneratedColumn()
-  public id!: number;
+  public id!: string;
```
Since the db entry for this is:
https://github.com/open-sauced/api/blob/2d08206998303ce31b4990b9ca91dd9239a8734b/migrations/20230905145301_add_user_list_contributors.sql#L4
with a uuid string, changed it to reflect in the type ORM as well

jpmcb commented on src/user-lists/entities/user-list-contributor.entity.ts:
```diff
@@ -15,15 +16,15 @@ import {
 import { DbUser } from "../../user/user.entity";
 
-@Entity({ name: "user_organizations" })
+@Entity({ name: "user_list_contributors" })
```
I'm assuming this was a copy paste error? I was having weird behavior where these columns were getting put into the `user_organizations` table. I'm assuming since the table name is:

https://github.com/open-sauced/api/blob/2d08206998303ce31b4990b9ca91dd9239a8734b/migrations/20230905145301_add_user_list_contributors.sql#L1

that it should also be reflected here

brandonroberts replied:
🤦  Yep!

jpmcb commented on src/user-lists/entities/user-list.entity.ts:
```diff
@@ -19,7 +19,7 @@ import { ApiHideProperty } from "@nestjs/swagger";
 import { DbUser } from "../../user/user.entity";
 
-@Entity({ name: "insights" })
+@Entity({ name: "user_lists" })
```
Same here: seemed that the table name "insights" wasn't right

https://github.com/open-sauced/api/blob/2d08206998303ce31b4990b9ca91dd9239a8734b/migrations/20230905145101_add_user_lists.sql#L1

jpmcb commented on src/user-lists/user-list.controller.ts:
```diff
@@ -10,18 +10,40 @@ import {
 export class UserListController {
   constructor(private readonly userListService: UserListService) {}
 
+  @Get("/")
+  @ApiOperation({
+    operationId: "getListsForUser",
+    summary: "Gets lists for the authenticated user",
+  })
```
Using this:

```sh
curl -X 'GET' \
  'http://localhost:3001/v1/lists?page=1&limit=10&range=30' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer <redacted>'
```

returns the two lists I made during playing around with this in the beta env
```json
{
  "data": [
    {
      "id": "17e0f1c2-c556-4734-bfc4-86d43efc4087",
      "name": "My real list",
      "is_public": false,
      "created_at": "2023-09-07T22:42:30.680Z",
      "updated_at": "2023-09-07T22:42:30.680Z",
      "deleted_at": null
    },
    {
      "id": "43fdd609-4936-4ff6-8083-199f46871262",
      "name": "John's tmp List",
      "is_public": false,
      "created_at": "2023-09-07T22:30:10.167Z",
      "updated_at": "2023-09-07T22:30:10.167Z",
      "deleted_at": null
    }
  ],
  "meta": {
    "page": 1,
    "limit": 10,
    "itemCount": 2,
    "pageCount": 1,
    "hasPreviousPage": false,
    "hasNextPage": false
  }
}
```

jpmcb commented on src/user-lists/user-list.service.ts:
```diff
@@ -82,11 +83,51 @@ export class UserListService {
+  async findContributorsByListId(pageOptionsDto: PageOptionsDto, listId: string) {
+    const queryBuilder = this.userListContributorRepository.createQueryBuilder("user_list_contributors");
+
+    queryBuilder
+      .innerJoin("users", "users", "user_list_contributors.user_id=users.id")
+      .addSelect("users.login", "user_lists_login")
+      .where("user_list_contributors.list_id = :listId", { listId });
```
@brandonroberts I'm confused how to get the `user_lists_login` to appear in the results:

```sh
curl -X 'GET' \
  'http://localhost:3001/v1/lists/17e0f1c2-c556-4734-bfc4-86d43efc4087/contributors?page=1&limit=10&range=30' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer <redacted>'
```

gets me

```json
{
  "data": [
    {
      "id": "72211ba0-8036-446d-98af-f5436ba11f4f",
      "user_id": 42211,
      "list_id": "17e0f1c2-c556-4734-bfc4-86d43efc4087",
      "created_at": "2023-09-07T22:42:30.793Z"
    },
    {
      "id": "c6f317ca-fadb-4ca6-b3a7-1a5686cc4ec5",
      "user_id": 23109390,
      "list_id": "17e0f1c2-c556-4734-bfc4-86d43efc4087",
      "created_at": "2023-09-07T22:42:31.002Z"
    }
  ],
  "meta": {
    "page": 1,
    "limit": 10,
    "itemCount": 2,
    "pageCount": 1,
    "hasPreviousPage": false,
    "hasNextPage": false
  }
}
```

which are the two users I added. But it's missing the user's login. I'm guessing it's because this route expects the `DbUserListContributors` back and is squashing extra data not apart of that type. Any ideas or other places we tack on data?

brandonroberts replied:
In order to see the derived field, it must be on the entity also. Here is an example of a field that's on the entity but. not in the database.

https://github.com/open-sauced/api/blob/beta/src/user/entities/user-highlight.entity.ts#L125-L134

In this case the entity field would be `login` and the alias would be `user_list_contributors_login` because TypeORM uses the entity table name as a prefix on each property in the entity.

jpmcb reviewed (commented):
@brandonroberts - this is ready for another review 💪🏼 let me know if there's anything that's not making sense! Tested this out on the dev database and is looking good!

jpmcb commented:
A few changes with the last force push:

- Added `FilterListContributorsDto` which is used as to filter users for the `v1/lists/contributors` get endpoint
  - This endpoint now accepts the Dto filter for "location", "timezone", "pr_velocity". The PR velocity query uses the last 30 days of PRs for a given user and rejects users who have a PR velocity of 0 (since this indicates they have no PRs). Feedback on this bit might be useful. In my testing I found this query to be _abit_ gnarly since it's perform a pretty broad select on potentially a large number of users (depending on how the other filters are configured). I wonder if a better way would be for us to have a field in the database for the `users` table that denotes PR velocity instead of trying to calculate it at runtime on each query.
  - Did quite abit of testing and it all seems to work well. You'll also notice I dropped the `INTEGER` bit of the PR velocity query since this would drop decimal precision for users who have really fast PR velocity (like Brandon's is in the `0.6` range) and ignore them with the `!= 0` clause.
- Adds the derived property for the `user_list_contributors_login`

brandonroberts commented:
@jpmcb I think having a regularly updated field like `recent_pr_velocity` would be good also. If we're going to have multiple time ranges, we'll have to keep it dynamic and make sure the DB indexes are keeping queries efficient. 

I think its a good step for these endpoints!

OgDev-01 reviewed (approved):
LGTM
